lib/util/comments.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * @fileoverview Utility functions for comments handling.
+ */
+'use strict';
+
+/**
+ * Backports sourceCode.getCommentsBefore() for ESLint 3
+ *
+ * @param {Object} nodeOrToken Node or token to get comments for.
+ * @param {Object} sourceCode The SourceCode object.
+ * @returns {Array} Array of comment tokens.
+ */
+function getCommentsBefore(nodeOrToken, sourceCode) {
+  const token = sourceCode.getFirstToken(nodeOrToken, {includeComments: true});
+  let previousComments = [];
+
+  // ESLint >=4.x
+  if (sourceCode.getCommentsBefore) {
+    previousComments = sourceCode.getCommentsBefore(token);
+  // ESLint 3.x
+  } else {
+    let currentToken = token;
+    do {
+      const previousToken = sourceCode.getTokenBefore(currentToken);
```

Comment:
Note: this code was extracted from `jsx-curly-spacing` and for some reason this line (or the next) returns `undefined` on ESLint 3.0.0 😕 Works well on 3.19.0.

Reply:
If we backported the eslint API, would that work?

Reply:
Looks like [both `getLocFromIndex` and `getIndexFromLoc`](https://github.com/eslint/eslint/blob/da9174e0798c1d785ddabb3ae405860fc5b89311/lib/util/source-code.js#L434-L504) are relying heavily on `lineStartIndices` and `lines` in the `SourceCode` class that get populated once in the constructor. The ESLint docs explicitly say [we can use `lines` (bottom of the section)](https://eslint.org/docs/developer-guide/working-with-rules#contextgetsourcecode) but there's no word about `lineStartIndices`. Potentially we can just copy over those functions, but the bigger issue with `getTokenBefore`/`getTokenAfter` remains, and those don't seem possible to be backported.

Reply:
Is there a way we could runtime-detect these APIs, and provide a reasonable degraded experience for older eslint users?

Reply:
I'm skeptical (not that we could detect them, but that there is a reasonable experience beyond just not applying fixes), but let me look into that.

Reply:
yeah that's what i meant - applying fewer fixes when they weren't good enough.